gr-digital/lib/decision_feedback_equalizer_impl.cc

```diff
@@ -0,0 +1,267 @@
+    bool adapt_after_training,
+    std::vector<gr_complex> training_sequence,
+    const std::string& training_start_tag)
+    : gr::sync_decimator("decision_feedback_equalizer",
+                         io_signature::make(1, 1, sizeof(gr_complex)),
+                         io_signature::makev(1,
+                                             3,
+                                             std::vector<int>{ sizeof(gr_complex),
+                                                               (int)((num_taps_forward +
+                                                                      num_taps_feedback) *
+                                                                     sizeof(gr_complex)),
+                                                               sizeof(unsigned short) }),
+                         sps),
+      filter::kernel::fir_filter_ccc(
+          sps,
+          vector<gr_complex>(num_taps_forward + num_taps_feedback, gr_complex(0, 0))),
+      d_num_taps_fwd(num_taps_forward),
+      d_num_taps_rev(num_taps_feedback),
+      d_sps(sps),
+      d_alg(alg),
+      d_adapt_after_training(adapt_after_training),
+      d_training_sequence(training_sequence),
+      d_training_start_tag(training_start_tag),
+      d_num_taps(num_taps_forward + num_taps_feedback),
+      d_decision_history(num_taps_feedback),
+      d_new_taps(num_taps_forward + num_taps_feedback),
+      d_updated(false),
+      d_training_sample(0),
+      d_filt_buf(num_taps_forward + num_taps_feedback)
+{
+    // not reversed at this point
+    d_new_taps[0] = gr_complex(1.0, 0.0); // help things along with initialized taps
+
+    if (training_start_tag == "" || training_sequence.empty()) {
+        d_training_state = equalizer_state_t::DD;
+    } else {
+        d_training_state = equalizer_state_t::IDLE;
+    }
+
+    set_history(d_num_taps_fwd);
+
+    alg->initialize_taps(d_new_taps);
+    // NOTE: the filter kernel reverses the taps internally
+    filter::kernel::fir_filter_ccc::set_taps(d_new_taps);
+}
+
+void decision_feedback_equalizer_impl::update_decision_history(gr_complex decision)
+{
+    std::rotate(d_decision_history.begin(),
+                d_decision_history.begin() + 1,
+                d_decision_history.end());
+    d_decision_history[d_decision_history.size() - 1] = decision;
+}
+
+int decision_feedback_equalizer_impl::work(int noutput_items,
+                                           gr_vector_const_void_star& input_items,
+                                           gr_vector_void_star& output_items)
+{
+
+    auto in = static_cast<const gr_complex*>(input_items[0]);
+    auto out = static_cast<gr_complex*>(output_items[0]);
+
+    int outlen = output_items.size();
+
+    unsigned short* state = nullptr;
+    gr_complex* taps = nullptr;
+
+    if (outlen > 1)
+        taps = static_cast<gr_complex*>(output_items[1]);
+    if (outlen > 2)
+        state = static_cast<unsigned short*>(output_items[2]);
+
+    if (d_updated) {
```

Comment:
`d_updated` is here read without holding a lock, thus is technically a data race and therefore undefined behaviour, I believe.

Maybe switch type to `std::atomic_flag`?

Reply:
Not sure about std::atomic_flag - looks like it might complicate the code, but I will protect d_updated

Reply:
Yeah atomics can be subtle, so usually it's better to use a mutex.